jobs/mysql_restore_job.go

```diff
@@ -0,0 +1,397 @@
+	if err != nil {
+		return 0, 0, errors.WithStack(err)
+	}
+
+	gid, err := strconv.Atoi(g.Gid)
+	if err != nil {
+		return 0, 0, errors.WithStack(err)
+	}
+
+	return uid, gid, nil
+}
+
+func isPathExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	switch {
+	case err == nil:
+		return true, nil
+	case os.IsNotExist(err):
+		return false, nil
+	default:
+		return false, errors.WithStack(err)
+	}
+}
+
+func restoreBackup(ctx context.Context, backupDirectory, mySQLDirectory string) error {
+	if output, err := exec.CommandContext( //nolint:gosec
+		ctx,
+		xtrabackupBin,
+		"--prepare",
+		"--target-dir="+backupDirectory,
+	).CombinedOutput(); err != nil {
+		return errors.Wrapf(err, "failed to prepare, output: %s", string(output))
+	}
+
+	exists, err := isPathExists(mySQLDirectory)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if exists {
+		postfix := ".old" + strconv.FormatInt(time.Now().Unix(), 10)
+		if err := os.Rename(mySQLDirectory, mySQLDirectory+postfix); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	if output, err := exec.CommandContext( //nolint:gosec
+		ctx,
+		xtrabackupBin,
+		"--copy-back",
+		"--datadir="+mySQLDirectory,
+		"--target-dir="+backupDirectory).CombinedOutput(); err != nil {
+		return errors.Wrapf(err, "failed to copy back, output: %s", string(output))
+	}
+
+	uid, gid, err := mySQLUserAndGroupIDs()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if err := chownRecursive(mySQLDirectory, uid, gid); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+// Run executes backup restore steps.
+func (j *MySQLRestoreJob) Run(ctx context.Context, send Send) (rerr error) {
+	j.l.Info("MySQL restore started")
+	defer func(start time.Time) {
+		entry := j.l.WithField("duration", time.Since(start).String())
+
+		if rerr != nil {
+			entry.Error("MySQL restore finished with error")
+		} else {
+			entry.Info("MySQL restore finished")
+		}
+	}(time.Now())
+
+	if j.location.S3Config == nil {
+		return errors.New("S3 config is not set")
+	}
+
+	if err := binariesInstalled(); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if _, _, err := mySQLUserAndGroupIDs(); err != nil {
+		return errors.WithStack(err)
+	}
+
+	tmpDir, err := ioutil.TempDir("", "backup-restore")
+	if err != nil {
+		return errors.Wrap(err, "cannot create temporary directory")
+	}
+	defer func() {
+		if err := os.RemoveAll(tmpDir); err != nil {
+			j.l.WithError(err).Error("failed to remove temporary directory")
+		}
+	}()
+
+	if err := restoreMySQLFromS3(ctx, j.name, &j.location, tmpDir); err != nil {
+		return errors.WithStack(err)
+	}
+
+	active, err := mySQLActive(ctx)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if active {
+		if err := stopMySQL(ctx); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	if err := restoreBackup(ctx, tmpDir, mySQLDirectory); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := startMySQL(ctx); err != nil {
```

Comment:
should we start MySQL if it wasn't active?

Reply:
good question, need to ask products about that

Reply:
From a technical standpoint, I need to start MySQL to ensure that the restored db files don't break MySQL - e.g., when I try to restore MySQL backup on top of the Percona Distribution for MySQL the last will fail to start.